Ensure dag.test uses serialized dag for testing

ephraimbuddy opened the pull request:
While `dag test` command uses serialized dag, dag.test was using
in-memory serialized dag making direct usage of dag.test method
resulting in error.

This PR fixes this and ensures dag.test parses dag if the dag is not
parsed

Closes: https://github.com/apache/airflow/issues/56657

ashb commented on airflow-core/src/airflow/serialization/serialized_objects.py:
```diff
@@ -2346,7 +2346,7 @@ def _create_orm_dagrun(
     # create the associated task instances
     # state is None at the moment of creation
-    run.verify_integrity(session=session, dag_version_id=dag_version.id)
+    run.verify_integrity(session=session, dag_version_id=dag_version.id if dag_version else None)
```
Do we need to run verify integrity at all if dag version is none? (I dont have the context loaded in my head right now)

ephraimbuddy replied:
Yes. We need to create the Task Instances

tatiana reviewed (commented):
@ephraimbuddy Thanks a lot for working on this so quickly.
Would it be worth to update the PR description to mention it closes #56657?

ashb commented on airflow-core/src/airflow/serialization/serialized_objects.py:
```diff
@@ -2318,7 +2318,7 @@ def _create_orm_dagrun(
         )
     dag_version = DagVersion.get_latest_version(dag.dag_id, session=session)
-    if not dag_version:
+    if not dag_version and triggering_user_name != "dag_test":
```
I guess this works, but i don't line this method of detection.

First question: why do we need a orm dag object anyway? (I think this is since that is where the scheduling logic is?)

Second question: since we know we aren't going to store anything here in the db: do we need to call this function, or would a specialised one that just creates the orm objects in memory fir the purposes of dag.test be better suited?

ephraimbuddy replied:
> First question: why do we need a orm dag object anyway? (I think this is since that is where the scheduling logic is?)


The ORM object here is used to ensure the dag version is linked to the created dag run and task instances. Since we don't need that in test, we can use the serdag in the test funtion itself:
https://github.com/apache/airflow/blob/947f4b980137d75530a60edf8af91e38074da2ae/task-sdk/src/airflow/sdk/definitions/dag.py#L1200-L1210

> Second question: since we know we aren't going to store anything here in the db: do we need to call this function, or would a specialised one that just creates the orm objects in memory fir the purposes of dag.test be better suited?

I think the serdag used in the dag.test method is enough. I can prevent calling the dag_version if triggering_user_name is dag_test. The dag_version is only used to ensure dagrun and created TI is linked to a version but tests can skip it

kaxil replied:
Wouldn't a better thing to detect would be to use `triggered_by` field with `DagRunTriggeredByType.TEST` since it is enum as opposed to raw string anyway

uranusjr commented on airflow-core/src/airflow/api_fastapi/execution_api/datamodels/taskinstance.py:
```diff
@@ -247,7 +247,7 @@ class TaskInstance(BaseModel):
     run_id: str
     try_number: int
-    dag_version_id: uuid.UUID
+    dag_version_id: uuid.UUID | None = None
```
```suggestion
    dag_version_id: uuid.UUID | None
```

I feel we don’t really need the default, but can just explicitly provide None when needed. Also it’d be a good idea to document somewhere (a comment?) when None is expected.

ephraimbuddy replied:
The TI.dag_version_id is nullable. We only enforce TI.dag_version_id on the constructor which can still create a TI even when the value of dag_version_id is null. So I think we don't need comment for this. However, let me know again if we should and I will add it

kaxil commented on airflow-core/src/airflow/serialization/serialized_objects.py:
```diff
@@ -2346,7 +2346,7 @@ def _create_orm_dagrun(
     # create the associated task instances
     # state is None at the moment of creation
-    run.verify_integrity(session=session, dag_version_id=dag_version.id)
+    run.verify_integrity(session=session, dag_version_id=dag_version.id if dag_version else None)
```
Is this not failing MyPy? `dag_version_id` (for `verify_integrity`) has a `UUIDType` only so technically does not accept None

ephraimbuddy replied:
It's a keyword-only params, so passes mypy due to that.

kaxil replied:
its not about keyword-only or positional arg though, the typehint says `UUIDType` so technically `None` shouldn't be allowed (both for static type-checking and just for general understanding)

kaxil reviewed (commented):
Fundamental question: Should `dag.test()` create a temporary serialized DAG version like it was doing in one of the early 3.0.x version, or should it bypass the serialization requirement entirely?

kaxil commented on airflow-core/tests/unit/cli/commands/test_dag_command.py:
```diff
@@ -661,7 +661,12 @@ def test_dag_test(self, mock_get_dag):
 
```
There needs to be a explicit test for ensuring dag.test does not error with serialized dag not found -- since I think this is 2nd or 3rd time we have regressed

kaxil replied:
Actually `airflow dags test` CLI works fine without serialized dags. It is just `dag.test` that doesn't work. So the test should be somewhere else not in here.

ephraimbuddy committed:
Ensure dag.test uses serialized dag for testing

While `dag test` command uses serialized dag, dag.test was using
in-memory serialized dag making direct usage of dag.test method
resulting in error.

This PR fixes this and ensures dag.test parses dag if the dag is not
parsed

ephraimbuddy committed:
fixup! Ensure dag.test uses serialized dag for testing

kaxil commented on airflow-core/tests/unit/models/test_dag.py:
```diff
@@ -179,6 +180,29 @@ def setup_method(self) -> None:
+        """
+        DAG.test() should auto-parse and sync the DAG if it's not serialized yet.
+        """
+        from airflow.models.dagbag import DBDagBag
```
Can go at the top of the file

kaxil commented on airflow-core/tests/unit/models/test_dag.py:
```diff
@@ -179,6 +180,29 @@ def setup_method(self) -> None:
+
+        # Ensure not serialized yet
+        assert DBDagBag().get_latest_version_of_dag(dag_id, session=session) is None
+        assert session.query(DagRun).filter(DagRun.dag_id == dag_id).scalar() is None
```
Isn't this old sqla syntax?

kaxil commented on airflow-core/tests/unit/models/test_dag.py:
```diff
@@ -179,6 +180,29 @@ def setup_method(self) -> None:
+        assert DBDagBag().get_latest_version_of_dag(dag_id, session=session) is None
+        assert session.query(DagRun).filter(DagRun.dag_id == dag_id).scalar() is None
+
+        dag = DAG(dag_id=dag_id, schedule=None)
```
```suggestion
        dag = DAG(dag_id=dag_id)
```

kaxil replied:
also worth adding a single hello world style task

kaxil commented on task-sdk/src/airflow/sdk/definitions/dag.py:
```diff
@@ -1197,17 +1197,37 @@ def test(
+            scheduler_dag = DBDagBag().get_latest_version_of_dag(dag_id=self.dag_id, session=session)
+            if not scheduler_dag:
+                from airflow.dag_processing.bundles.manager import DagBundlesManager
+                from airflow.dag_processing.dagbag import DagBag, sync_bag_to_db
```
This (`airflow.dag_processing.dagbag`) won't be present if we want to cherry-pick it in 3.1.1 -- it is `airflow.models.dagbag` in 3.1.1 -- maybe add try..except

ephraimbuddy replied:
I wonder if it would be better to keep it the way it was before and during cherrypicking, remove the import and do dagbag.sync_bag_to_db there?

kaxil replied:
As long as you can create a manual backport PR and test it in v3-1-test I am fine with that approach :)

ephraimbuddy committed:
Update query to v2 and do backcompat for 3.1

kaxil reviewed (approved):
minor nitpicks -- lgtm otherwise

ephraimbuddy committed:
Remove backcompat

ephraimbuddy commented on task-sdk/src/airflow/sdk/definitions/dag.py:
```diff
@@ -1197,17 +1197,39 @@ def test(
+                            dag_folder=bundle.path, bundle_path=bundle.path, include_examples=False
+                        )
+                        sync_bag_to_db(dagbag, bundle.name, bundle.version)
+                        # type: ignore[attr-defined]
```
```suggestion
```

ephraimbuddy committed:
Return previous behaviour with SerializedDag

ephraimbuddy commented on task-sdk/src/airflow/sdk/definitions/dag.py:
```diff
@@ -1197,7 +1197,33 @@ def test(
+                    version = DagVersion.get_version(self.dag_id)
+                    if version:
+                        break
+            scheduler_dag = SerializedDAG.deserialize_dag(SerializedDAG.serialize_dag(self))
```
Surprised that using scheduler_dag instead of this, makes some tests fail. I wonder if those tests are doing the correct thing
